File: thurstone/clustering.py

from __future__ import annotations
from dataclasses import dataclass
from typing import List, Sequence, Tuple
import numpy as np
from .density import Density
from .lattice import UniformLattice

def _int_centered(offsets: Sequence[float]) -> List[float]:
    finite = [o for o in offsets if np.isfinite(o)]
    if not finite:
        return list(offsets)
    mean_int = int(np.mean(finite))
    return [o - mean_int for o in offsets]

def _divide_offsets(centered_offsets: Sequence[float]) -> float:
    """Choose a divider that maximizes the biggest gap near the center; ensures two non-empty groups."""
    srt = sorted(centered_offsets)
    if len(srt) <= 2:
        return float(np.mean(srt))
    gaps = np.diff(srt)
    idx = int(np.argmax(np.abs(gaps)))
    return float(0.5*(srt[idx] + srt[idx+1]))

@dataclass
class ClusterSplitter:
    unit_ratio: float = 3.0
    max_depth: int = 3

    def extended_state_prices(self, base: Density, offsets: Sequence[float]) -> List[float]:
        """Offsets may include +/-inf; returns normalized winning probabilities."""
        n = len(offsets)
        if n == 1:
            return [1.0]
        # handle +inf (no chance)
        pos_inf_idx = [i for i,o in enumerate(offsets) if o == float('inf')]
        if pos_inf_idx:
            finite_idx = [i for i in range(n) if i not in pos_inf_idx]
            if not finite_idx:
                return [1.0/n for _ in range(n)]
            finite_prices = self.extended_state_prices(base, _int_centered([offsets[i] for i in finite_idx]))
            out = [0.0 for _ in range(n)]
            for j, idx in enumerate(finite_idx):
                out[idx] = finite_prices[j]
            return out
        # handle -inf (certain winners share)
        neg_inf_idx = [i for i,o in enumerate(offsets) if o == float('-inf')]
        if neg_inf_idx:
            p = [0.0 for _ in range(n)]
            share = 1.0/len(neg_inf_idx)
            for idx in neg_inf_idx:
                p[idx] = share
            return p

        L = base.lattice.L
        W = int(base.approx_support_width())
        centered = _int_centered(offsets)

        # boundaries
        lower_bound = -L + W
        upper_bound = L - W
        hang_left = [i for i,o in enumerate(centered) if o < lower_bound]
        hang_right = [i for i,o in enumerate(centered) if o > upper_bound]
        # If all contestants hang on the same side and are tightly bunched relative to support,
        # treat them as indistinguishable at this resolution (equal share).
        if len(hang_left) == n or len(hang_right) == n:
            spread = float(max(centered) - min(centered)) if n > 0 else 0.0
            if spread < W:
                return [1.0/n for _ in range(n)]
        if (not hang_left) and (not hang_right):
            from .inference import densities_from_offsets, state_prices_from_densities
            dens = densities_from_offsets(base, centered)
            return state_prices_from_densities(dens)

        # stop recursion
        if self.max_depth <= 0:
            for i in hang_right:
                centered[i] = float('inf')
            for i in hang_left:
                centered[i] = float('-inf')
            return self.extended_state_prices(base, centered)

        # split symmetrically
        divider = _divide_offsets(centered)
        left_idx  = [i for i,o in enumerate(centered) if o < divider]
        right_idx = [i for i,o in enumerate(centered) if o >= divider]
        if len(left_idx)==0 or len(right_idx)==0:
            if len(left_idx)==0:
                left_idx = [int(np.argmin(centered))]
                right_idx = [i for i in range(n) if i not in left_idx]
            else:
                right_idx = [int(np.argmax(centered))]
                left_idx = [i for i in range(n) if i not in right_idx]

        # coarse group shares
        dilated = base.dilate(unit_ratio=self.unit_ratio)
        dil_offsets = [o / self.unit_ratio for o in centered]
        from .inference import densities_from_offsets, state_prices_from_densities
        coarse_prices = state_prices_from_densities(densities_from_offsets(dilated, dil_offsets))
        left_share  = float(sum(coarse_prices[i] for i in left_idx))
        right_share = 1.0 - left_share

        # refine inside groups
        left_prices_rel  = self.__class__(self.unit_ratio, self.max_depth - 1).extended_state_prices(base, [centered[i] for i in left_idx])
        right_prices_rel = self.__class__(self.unit_ratio, self.max_depth - 1).extended_state_prices(base, [centered[i] for i in right_idx])

        out = [0.0 for _ in range(n)]
        for j, idx in enumerate(left_idx):
            out[idx] = left_share * left_prices_rel[j]
        for j, idx in enumerate(right_idx):
            out[idx] = right_share * right_prices_rel[j]

        S = sum(out)
        if S <= 0:
            raise ValueError("Extended state prices have non-positive total mass.")
        assert 0.999 <= S <= 1.001, f"State prices not normalized in extended offsets; sum={S}"
        return [oi / S for oi in out]

File: docs/js/thurstone/clustering.js
// Port of thurstone/clustering.py
// ClusterSplitter handles offsets that hang off the lattice (walkovers, equal-share).

import { Density } from "./density.js";
import { Race } from "./pricing.js";
import { winnerOfMany } from "./order_stats.js";

function intCentered(offsets) {
  const finite = offsets.filter(o => Number.isFinite(o));
  if (finite.length === 0) return [...offsets];
  const meanInt = Math.trunc(finite.reduce((a, b) => a + b, 0) / finite.length);
  return offsets.map(o => o - meanInt);
}

function divideOffsets(centered) {
  const srt = [...centered].sort((a, b) => a - b);
  if (srt.length <= 2) return srt.reduce((a, b) => a + b, 0) / srt.length;
  let bestGap = -Infinity, bestIdx = 0;
  for (let i = 0; i < srt.length - 1; i++) {
    const g = Math.abs(srt[i + 1] - srt[i]);
    if (g > bestGap) { bestGap = g; bestIdx = i; }
  }
  return 0.5 * (srt[bestIdx] + srt[bestIdx + 1]);
}

function densitiesFromOffsets(base, offsets) {
  return offsets.map(o => base.shiftFractional(o));
}

function statePricesFromDensities(densities) {
  return Array.from(new Race(densities).statePrices());
}

export class ClusterSplitter {
  constructor({ unitRatio = 3.0, maxDepth = 3 } = {}) {
    this.unitRatio = unitRatio;
    this.maxDepth = maxDepth;
  }

  extendedStatePrices(base, offsets) {
    const n = offsets.length;
    if (n === 1) return [1.0];

    // Handle +inf (cannot win)
    const posInf = [];
    for (let i = 0; i < n; i++) if (offsets[i] === Infinity) posInf.push(i);
    if (posInf.length > 0) {
      const finiteIdx = [];
      for (let i = 0; i < n; i++) if (!posInf.includes(i)) finiteIdx.push(i);
      if (finiteIdx.length === 0) return Array(n).fill(1.0 / n);
      const fp = this.extendedStatePrices(base, intCentered(finiteIdx.map(i => offsets[i])));
      const out = Array(n).fill(0.0);
      for (let j = 0; j < finiteIdx.length; j++) out[finiteIdx[j]] = fp[j];
      return out;
    }
    // Handle -inf (certain winners share equally)
    const negInf = [];
    for (let i = 0; i < n; i++) if (offsets[i] === -Infinity) negInf.push(i);
    if (negInf.length > 0) {
      const p = Array(n).fill(0.0);
      const share = 1.0 / negInf.length;
      for (const i of negInf) p[i] = share;
      return p;
    }

    const L = base.lattice.L;
    const W = base.approxSupportWidth();
    const centered = intCentered(offsets);

    const lowerBound = -L + W;
    const upperBound = L - W;
    const hangLeft = [], hangRight = [];
    for (let i = 0; i < n; i++) {
      if (centered[i] < lowerBound) hangLeft.push(i);
      if (centered[i] > upperBound) hangRight.push(i);
    }
    if (hangLeft.length === 0 && hangRight.length === 0) {
      return statePricesFromDensities(densitiesFromOffsets(base, centered));
    }

    if (this.maxDepth <= 0) {
      for (const i of hangRight) centered[i] = Infinity;
      for (const i of hangLeft) centered[i] = -Infinity;
      return this.extendedStatePrices(base, centered);
    }

    const divider = divideOffsets(centered);
    let leftIdx = [], rightIdx = [];
    for (let i = 0; i < n; i++) (centered[i] < divider ? leftIdx : rightIdx).push(i);
    if (leftIdx.length === 0 || rightIdx.length === 0) {
      if (leftIdx.length === 0) {
        let am = 0; for (let i = 1; i < centered.length; i++) if (centered[i] < centered[am]) am = i;
        leftIdx = [am]; rightIdx = []; for (let i = 0; i < n; i++) if (i !== am) rightIdx.push(i);
      } else {
        let am = 0; for (let i = 1; i < centered.length; i++) if (centered[i] > centered[am]) am = i;
        rightIdx = [am]; leftIdx = []; for (let i = 0; i < n; i++) if (i !== am) leftIdx.push(i);
      }
    }

    const densLeft = densitiesFromOffsets(base, leftIdx.map(i => centered[i]));
    const densRight = densitiesFromOffsets(base, rightIdx.map(i => centered[i]));
    if (densLeft.length === 0 || densRight.length === 0) {
      return statePricesFromDensities(densitiesFromOffsets(base, centered));
    }
    const { density: repLeft } = winnerOfMany(densLeft);
    const { density: repRight } = winnerOfMany(densRight);
    const groupPrices = new Race([repLeft, repRight]).statePrices();
    const leftShare = groupPrices[0];
    const rightShare = groupPrices[1];

    let leftRel, rightRel;
    if (leftShare <= rightShare) {
      const sub = new ClusterSplitter({ unitRatio: this.unitRatio, maxDepth: this.maxDepth - 1 });
      leftRel = sub.extendedStatePrices(base, leftIdx.map(i => centered[i]));
      rightRel = statePricesFromDensities(densRight);
      const Sr = rightRel.reduce((a, b) => a + b, 0);
      if (Sr > 0) rightRel = rightRel.map(p => p / Sr);
    } else {
      const sub = new ClusterSplitter({ unitRatio: this.unitRatio, maxDepth: this.maxDepth - 1 });
      rightRel = sub.extendedStatePrices(base, rightIdx.map(i => centered[i]));
      leftRel = statePricesFromDensities(densLeft);
      const Sl = leftRel.reduce((a, b) => a + b, 0);
      if (Sl > 0) leftRel = leftRel.map(p => p / Sl);
    }

    const out = Array(n).fill(0.0);
    for (let j = 0; j < leftIdx.length; j++) out[leftIdx[j]] = leftShare * leftRel[j];
    for (let j = 0; j < rightIdx.length; j++) out[rightIdx[j]] = rightShare * rightRel[j];

    const S = out.reduce((a, b) => a + b, 0);
    if (S <= 0) throw new Error("Extended state prices have non-positive total mass.");
    return out.map(o => o / S);
  }
}
